src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioErrorReportingService.ExceptionFormatting.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    internal partial class VisualStudioErrorReportingService
+    {
+        private static string GetFormattedExceptionStack(Exception exception)
+        {
+            var aggregate = exception as AggregateException;
+            if (aggregate != null)
+            {
+                return GetStackForAggregateException(exception, aggregate);
+            }
+
+            return GetStackForException(exception, false);
+        }
+
+        private static string GetStackForAggregateException(Exception exception, AggregateException aggregate)
+        {
+            var text = GetStackForException(exception, true);
+            for (int i = 0; i < aggregate.InnerExceptions.Count; i++)
+            {
+                text = string.Format("{0}{1}---> (Inner Exception #{2}) {3}{4}{5}", text,
+                    Environment.NewLine, i, GetFormattedExceptionStack(aggregate.InnerExceptions[i]), "<---", Environment.NewLine);
+            }
+
+            return text;
+        }
+
+        private static string GetStackForException(Exception exception, bool includeMessageOnly)
+        {
+            var message = exception.Message;
+            var className = exception.GetType().ToString();
+            var stackText = message.Length <= 0
+                ? className
+                : className + " : " + message;
+            var innerException = exception.InnerException;
+            if (innerException != null)
+            {
+                if (includeMessageOnly)
+                {
+                    do
+                    {
+                        stackText += " ---> " + innerException.Message;
+                        innerException = innerException.InnerException;
+                    } while (innerException != null);
+                }
+                else
+                {
+                    stackText += " ---> " + GetFormattedExceptionStack(innerException) + Environment.NewLine +
+                                 "   " + ServicesVSResources.End_of_inner_exception_stack;
+                }
+            }
+            
+            return stackText + Environment.NewLine + GetAsyncStackTrace(exception);
+        }
+
+        private static string GetAsyncStackTrace(Exception exception)
+        {
+            var stackTrace = new StackTrace(exception);
+
+            var stackFrames = stackTrace.GetFrames();
+            if (stackFrames == null)
+            {
+                return string.Empty;
+            }
+
+            var firstFrame = true;
```

Comment:
any time i see something lk ethis, i feel like it should be replaced with a string.Join(Environment.NewLine, values)